spec/formatting.md

```diff
@@ -18,3 +18,71 @@ the local variable takes precedence.
 
 It is an error for a local variable definition to
 refer to a local variable that's defined after it in the message.
+
+## Error Handling
+
+During the formatting of a message,
+various errors may be encountered.
+These are divided to the following categories:
+
+- **Syntax errors** occur when the syntax representation of a message is invalid.
+- **Resolution errors** occur when the runtime value of a part of a message
+  cannot be determined.
+
+  - **Unresolved Variable** errors occur when a variable reference cannot be resolved.
+
+- **Selection errors** cover failures encountered during selection.
+
+  - **Selector errors** are failures in the matching of a key to a specific selector.
+  - **Missing Fallback** errors occur when no Variant is selected
+    due to the message not including a Variant with only catch-all keys.
+
+- **Formatting errors** occur during the formatting of a resolved value,
+  for example when encountering a value with an unsupported type
+  or an internally inconsistent set of options.
+
+During selection, an expression handler must only emit Resolution and Selection errors.
+During formatting, an expression handler must only emit Resolution and Formatting errors.
+
+In all cases, when encountering an error,
+a message formatter must provide some representation of the message.
+An informative error or errors must also be separately provided.
+
+When an error occurs in the syntax or resolution of an Expression or MarkupStart Option,
+the Expression or MarkupStart in question is processed as if the option was not defined.
+This may allow for the fallback handling described below to be avoided,
+though an error must still be emitted.
+
+When an error occurs within a Selector,
+the selector must not match any VariantKey other than the catch-all `*`
+and a Selector error is emitted.
+When selection fails to match any Variant,
+an empty string is used as the formatted string representation of the message
+and a Missing Fallback error is emitted.
+
+When an error occurs in a Placeholder that is being formatted,
+the fallback string representation of the Placeholder
+always starts with U+007B LEFT CURLY BRACKET `{`
+and ends with U+007D RIGHT CURLY BRACKET `}`.
+Between the brackets, the following contents are used:
+
+- Expression with Literal Operand: U+0028 LEFT PARENTHESIS `(`
+  followed by the value of the Literal,
+  and then by U+0029 RIGHT PARENTHESIS `)`
+- Expression with Variable Operand: U+0024 DOLLAR SIGN `$`
+  followed by the Variable Name of the Operand
+- Expression with no Operand: U+003A COLON `:` followed by the Expression Name
+- Markup start: U+002B PLUS SIGN `+` followed by the MarkupStart Name
+- Markup end: U+002D HYPHEN-MINUS `-` followed by the MarkupEnd Name
+- Otherwise: Three U+003F QUESTION MARK `?` characters, i.e. `???`
```

Comment:
Should we consider FFFC or FFFD here? Black diamonds stand out more and are less likely to just "become data", whereas question marks can appear in real data?

Reply:
```suggestion
- Otherwise: Three U+FFFD REPLACEMENT CHARACTER `�` characters, i.e. `���`
```

I definitely like that better than a `?`. Should the repetition be removed?

Reply:
I think removing the repetition makes sense.

Reply:
```suggestion
- Otherwise: The U+FFFD REPLACEMENT CHARACTER `�` character
```